codec.go

```diff
@@ -0,0 +1,173 @@
+	// should return (nil, ErrUnsupportedSampleCodec) in the event c
+	// is not AnySampleCodec.
+	Encoder func(w io.WriteCloser, c sample.Codec) (sound.Sink, error)
+
+	// RandomAccess tries to turn an io.ReadWriteSeeker into sound.RandomAccess.
+	// If the codec does not make use of a defined sample.Codec and c is
+	// not AnySampleCodec, then the function should return (nil, ErrUnsupporteSampleCodec).
+	RandomAccess func(ws io.ReadWriteSeeker, c sample.Codec) (sound.RandomAccess, error)
+}
+
+var codecs []Codec
+
+// RegisterCodec registers a codec so that consumers of this package
+// can treat sound I/O generically and switch between codecs.
+//
+// A package "p" implementing a Codec can register a codec in its init()
+// function.
+//
+// Codecs registered by zikichombo.org have priority in the range [1000..2000).
+// Lower priority values are considered higher priority.
+//
+// Although c is a pointer, a "deep" copy of c is added to the list of registered codecs.
+func RegisterCodec(c *Codec) {
+	exts := make([]string, len(c.Extensions))
+	for i, ext := range c.Extensions {
+		exts[i] = ext
+	}
+
+	codecs = append(codecs, Codec{Priority: c.Priority,
+		Extensions:         exts,
+		Magic:              c.Magic,
+		DefaultSampleCodec: c.DefaultSampleCodec,
+		Decoder:            c.Decoder,
+		Encoder:            c.Encoder,
+		RandomAccess:       c.RandomAccess})
+}
+
+type sniffReader interface {
+	io.ReadCloser
+	Peek(int) ([]byte, error)
+}
+
+type sr struct {
+	rc io.ReadCloser
+	*bufio.Reader
+}
+
+func (r *sr) Close() error {
+	return r.Close()
+}
+
+func asSniffReader(r io.ReadCloser) sniffReader {
+	if sr, ok := r.(sniffReader); ok {
+		return sr
+	}
+	return &sr{
+		Reader: bufio.NewReader(r),
+		rc:     r}
+}
+
+// CodecFor tries to find a codec based on a filename extension.
+//
+// The returned codec, although a pointer to a struct with fields, should be
+// treated as read-only.  Not doing so may result in race conditions or worse.
+//
+// In case of conflict, the returned codec is a codec with lowest Priority
+// value.  In case of confict taking into account the priority, the first codec
+// from a call to RegisterCodec is returned.  As this might depend on package
+// initialisation order, it is recommended to Codec implementations intended
+// for library use use even valued priorities so a given application (with a
+// main()) may override package initialistion order by using an appropriate odd Priority
+// value.
+func CodecFor(ext string) (*Codec, error) {
+	minPriority := int(math.MaxInt32)
+	var bestCodec *Codec
+	for i := range codecs {
+		c := &codecs[i]
+		for _, codExt := range c.Extensions {
+			if ext == codExt {
+				if c.Priority < minPriority {
+					bestCodec = c
+				}
+			}
+		}
+	}
+	if bestCodec == nil {
+		return nil, ErrUnknownCodec
+	}
+	return bestCodec, nil
+}
+
+// Decoder
+func Decoder(r io.ReadCloser) (sound.Source, sample.Codec, error) {
+	return nil, sample.SFloat32L, nil
```

Comment:
Not yet implemented. Either return error, e.g. `errors.New("codec.Decoder: not yet implemented")` or implement ^^